cpp/include/rapidsmpf/streaming/core/channel.hpp

```diff
@@ -191,6 +193,169 @@ class Channel {
+    }
+
+    /**
+     * @brief Obtain a ticket to send a message.
+     *
+     * Suspends if all tickets are currently handed out.
+     *
+     * @return A coroutine producing a new ticket to be used in `send`.
+     */
+    [[nodiscard]] coro::task<Ticket> acquire() {
+        auto result = co_await semaphore_.acquire();
+        RAPIDSMPF_EXPECTS(
+            result == coro::semaphore_acquire_result::acquired, "Semaphore was shutdown"
+        );
+        co_return Ticket{};
+    }
+
+    /**
+     * @brief Release a ticket that has been converted to a sent message.
+     *
+     * @param receipt The receipt indicating one has sent a message.
+     * @return A coroutine representing completion of the release.
+
+     * @note To avoid immediately transferring control in this
+     * executing thread to any waiting tasks, one should `yield` in
+     * the executor before `release`ing.
+     *
+     * Here is a typical pattern:
+     * @code{.cpp}
+     * auto ticket = co_await channel.acquire();
+     * auto data = do_expensive_work();
+     * auto [_, receipt] = co_await channel.send(std::move(ticket));
+     * co_await executor->yield();
+     * co_await channel.release(std::move(receipt));
+     * @endcode
+     */
+    coro::task<void> release([[maybe_unused]] Receipt&& receipt) {
+        RAPIDSMPF_EXPECTS(!semaphore_.is_shutdown(), "Semaphore was shutdown");
+        co_await semaphore_.release();
+    }
+
+    /**
+     * @brief Asynchronously receive a message from the channel.
+     *
+     * Suspends if the channel is empty.
+     *
+     * @return A coroutine that evaluates to the message, which will be empty if the
+     * channel is shut down.
+     */
+    coro::task<Message> receive() {
+        auto msg = co_await rb_.consume();
+        if (msg.has_value()) {
+            co_return std::move(*msg);
+        } else {
+            co_return Message{};
+        }
+    }
+
+    /**
+     * @brief Drains all pending messages from the channel and shuts it down.
+     *
+     * This is intended to ensure all remaining messages are processed.
+     *
+     * @param executor The thread pool used to process remaining messages.
+     * @return A coroutine representing the completion of the shutdown drain.
+     */
+    Node drain(std::shared_ptr<coro::thread_pool> executor) {
+        return rb_.shutdown_drain(std::move(executor));
+    }
+
+    /**
+     * @brief Immediately shuts down the channel.
+     *
+     * Any pending or future send/receive operations will complete with failure.
+     *
+     * @return A coroutine representing the completion of the shutdown.
+     */
+    Node shutdown() {
+        return rb_.shutdown();
+    }
+
+    /**
+     * @brief Check whether the channel is empty.
+     *
+     * @return True if there are no messages in the buffer.
+     */
+    [[nodiscard]] bool empty() const noexcept {
+        return rb_.empty();
+    }
+
+  private:
+    coro::ring_buffer<Message, 1> rb_;
```

Comment:
I think the semaphore guarantees the size of container to hold messages. So, cant we use a `deque` instead of a `ring_buffer` with the semaphore size? It feels like the semaphore AND ring_buffer seems redundant.

Reply:
Not quite.

The semaphore says "how many people can queue up with data and suspend at channel.send()".

This is _independent_ from the capacity of the channel.

If I used a deque then I have no way of applying back-pressure if the consumer of the channel is not ready yet. I can just keep acquiring tickets, pushing things into the channel and then releasing tickets.